roachtest: sqlsmith/setup=seed/setting=no-mutations failed

roachtest.sqlsmith/setup=seed/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3838980&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3838980&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations) on master @ [260b7930e717c76bc421ae3f34e9a939fc896b50](https://github.com/cockroachdb/cockroach/commits/260b7930e717c76bc421ae3f34e9a939fc896b50):


```
																							AS tab_47872 (col_118482)
																				)
																		SELECT
																			'010500004000000000':::GEOMETRY AS col_118483
																		FROM
																			with_19749 AS cte_ref_5815
																		WHERE
																			false
																		LIMIT
																			1:::INT8
																	)
																),
																(
																	st_geomfromgeojson(e'\rc\\':::STRING::STRING):::GEOMETRY::GEOMETRY
																)
														)
															AS tab_47873 (col_118484)
												)
										SELECT
											B'010111000000101100110100101010110110010011' AS col_118485
										FROM
											with_19746 AS cte_ref_5816
										ORDER BY
											cte_ref_5816.col_118477 ASC, cte_ref_5816.col_118477 DESC
										LIMIT
											1:::INT8
									)
								)
						)
							AS tab_47874 (col_118486)
					INTERSECT
						SELECT
							*
						FROM
							(
								VALUES
									(B'011010011011111110110011000110100000001001'),
									(B'110111110111000111011000010100100000010100'),
									(B'011101011110101001111100010101100111111001')
							)
								AS tab_47875 (col_118487)
				)
		SELECT
			tab_47876._enum AS col_118488, cte_ref_5817.col_118473 AS col_118489
		FROM
			with_19744 AS cte_ref_5817, defaultdb.public.seed@seed__int8__float8__date_idx AS tab_47876
		ORDER BY
			cte_ref_5817.col_118472 ASC, tab_47876._int4 ASC, cte_ref_5817.col_118472 DESC
		LIMIT
			32:::INT8;
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=seed/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.sqlsmith/setup=seed/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3889443&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3889443&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations) on master @ [406e0fe1e44cabc8a3384703c642469567c6915e](https://github.com/cockroachdb/cockroach/commits/406e0fe1e44cabc8a3384703c642469567c6915e):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:224,sqlsmith.go:257,test_runner.go:779: error: pq: internal error: unable to vectorize execution plan: CASE WHEN "$9" THEN CASE WHEN "$11" NOT IN (e'O:2\x1f``\x13=', "$11", '') THEN ((),) ELSE (NULL:::RECORD,) END ELSE CAST(NULL AS RECORD) END: incompatible value type: incompatible value type: tuples [((),) (NULL:::RECORD,)] are not the same type: expected NULL:::RECORD to be of type tuple, found type unknown
		stmt:
		WITH
			with_11432 (col_68653, col_68654)
				AS (
					SELECT
						DISTINCT
						CASE
						WHEN tab_27762._bool
						THEN CASE
						WHEN tab_27762._string::STRING NOT IN (e'O:2\x1f``\x13=':::STRING, tab_27762._string, '':::STRING)
						THEN ((),)
						ELSE (NULL,)
						END
						ELSE NULL
						END
							AS col_68653,
						COALESCE('00:01:27.840854+07:17:00':::TIMETZ, NULL) AS col_68654
					FROM
						defaultdb.public.seed@[0] AS tab_27762
				)
		SELECT
			((),) AS col_68655, '09:16:57.250339-00:04:00':::TIMETZ AS col_68656
		FROM
			with_11432 AS cte_ref_3292
		ORDER BY
			cte_ref_3292.col_68654 DESC;
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=seed/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/catalog/typedesc/type_desc.go
+++ b/pkg/sql/catalog/typedesc/type_desc.go
@@ -741,31 +741,26 @@ func (desc *immutable) MakeTypesT(
 func EnsureTypeIsHydrated(
 	ctx context.Context, t *types.T, res catalog.TypeDescriptorResolver,
 ) error {
-	// maybeHydrateType checks if t is a user-defined type that hasn't been
-	// hydrated yet, and installs the metadata if so.
-	maybeHydrateType := func(ctx context.Context, t *types.T, res catalog.TypeDescriptorResolver) error {
-		if !t.UserDefined() || t.IsHydrated() {
-			return nil
-		}
-		id, err := GetUserDefinedTypeDescID(t)
-		if err != nil {
-			return err
-		}
-		elemTypName, elemTypDesc, err := res.GetTypeDescriptor(ctx, id)
-		if err != nil {
-			return err
-		}
-		return elemTypDesc.HydrateTypeInfoWithName(ctx, t, &elemTypName, res)
-	}
 	if t.Family() == types.TupleFamily {
 		for _, typ := range t.TupleContents() {
-			if err := maybeHydrateType(ctx, typ, res); err != nil {
+			if err := EnsureTypeIsHydrated(ctx, typ, res); err != nil {
 				return err
 			}
 		}
 		return nil
 	}
-	return maybeHydrateType(ctx, t, res)
+	if !t.UserDefined() || t.IsHydrated() {
+		return nil
+	}
+	id, err := GetUserDefinedTypeDescID(t)
+	if err != nil {
+		return err
+	}
+	elemTypName, elemTypDesc, err := res.GetTypeDescriptor(ctx, id)
+	if err != nil {
+		return err
+	}
+	return elemTypDesc.HydrateTypeInfoWithName(ctx, t, &elemTypName, res)
 }
 
 // HydrateTypesInTableDescriptor uses res to install metadata in the types

--- a/pkg/sql/logictest/testdata/logic_test/distsql_enum
+++ b/pkg/sql/logictest/testdata/logic_test/distsql_enum
@@ -1,0 +1,33 @@
+# LogicTest: 5node-default-configs
+
+# Regression test for nested tuple enum hydration (#74189)
+statement ok
+CREATE TYPE greeting AS ENUM ('hello')
+
+statement ok
+CREATE TABLE IF NOT EXISTS seed AS
+	SELECT
+		enum_range('hello'::greeting)[g] as _enum
+	FROM
+		generate_series(1, 1) AS g
+
+query TT nodeidx=3
+WITH w (col)
+				AS (
+					SELECT
+						*
+					FROM
+						(
+							VALUES
+								(
+									((('hello':::greeting, 0), 0))
+								)
+						)
+				)
+		SELECT
+			seed._enum, w.col
+		FROM
+			w, seed
+----
+hello     ("(hello,0)",0)
+

